File: client/src/models/craft/Recipe.ts
import { Item } from "../items/Item";
import { RecipeMaterial } from "./RecipeMaterial";

export class Recipe {
  constructor(
    name: string,
    type: string,
    item: Item,
    duration: number,
    recipeMaterials: RecipeMaterial[],
    instrumentType: string,
    instrumentRequiredLevel: number,
    instrumentExpectedLevel: number) {
    this.name = name;
    this.type = type;
    this.item = item;
    this.duration = duration;
    this.recipeMaterials = recipeMaterials;
    this.instrumentType = instrumentType;
    this.instrumentRequiredLevel = instrumentRequiredLevel;
    this.instrumentExpectedLevel = instrumentExpectedLevel;
  }

  name: string;
  type: string;
  item: Item;
  duration: number;
  recipeMaterials: RecipeMaterial[];
  instrumentType: string;
  instrumentRequiredLevel: number;
  instrumentExpectedLevel: number;
}

File: EntityModels.Postgresql/Recipe.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace EntityModels.Postgresql;

[Table("Recipe")]
public partial class Recipe
{
    [Key] public string Name { get; set; } = null!;
    [StringLength(50)] public string Type { get; set; } = null!;
    public string ItemName { get; set; } = null!;
    public Item Item { get; set; } = null!;
    public int DurationSeconds { get; set; }
    public ICollection<RecipeMaterial> RecipeMaterials { get; set; } = null!;
    [StringLength(50)] public string InstrumentType { get; set; } = null!;
    public int InstrumentRequiredLevel { get; set; }
    public int InstrumentExpectedLevel { get; set; } //  (x2 faster for level above and 4x slower for level below)
}